### ISSUE TITLE
GSON: For List<> field no JSON name is generated

### ISSUE THREAD

**Issue by pogo19** [immutables/immutables#369]
Using Immutables 2.2.7 with GSON, I have this class:

```java
@Immutable
@Gson.TypeAdapters
public abstract class ListContext {

	@Parameter
	@Nullable
	public abstract Long getStartIndex();

	@Parameter
	@Nullable
	public abstract Long getEndIndex();

	@Parameter
	@Nullable
	public abstract Set<Filter<Object>> getFilters();

	@Parameter
	@Nullable
	public abstract List<Sort> getSorts();

}
```

I also define `Filter<>` and `Sort` classes. Problem is that generated GSON TypeAdapter doesn't write name for `"sorts"`, not tested with non-null `"filters"` but it looks like the same problem. Maybe it is the combination of `@Nullable` and collection.

Generated `GsonAdaptersListContext`, part of `writeListContext(..)`:
```java
@Nullable Set<Filter<Object>> filtersElements = instance.getFilters();
      if (filtersElements != null) {
        out.beginArray();
        for (Filter<Object> e : filtersElements) {
          filtersTypeAdapter.write(out, e);
        }
        out.endArray();
      } else if (out.getSerializeNulls()) {
        out.name("filters");
        out.nullValue();
      }
      @Nullable List<Sort> sortsElements = instance.getSorts();
      if (sortsElements != null) {
        out.beginArray();
        for (Sort e : sortsElements) {
          sortsTypeAdapter.write(out, e);
        }
        out.endArray();
      } else if (out.getSerializeNulls()) {
        out.name("sorts");
        out.nullValue();
      }
      out.endObject();
```

**Comment by elucash** [comment]
Isn't this just a plain effect of Gson's standard `out.getSerializeNulls()` property that Immutables tries to honor? It is `false` by default allowing for prettier (or more concise) JSON.
you can set `serializeNulls` on the level of `JsonWriter` or `GsonBuilder`. Please, let me know if it's not this case.

**Comment by pogo19** [comment]
To clarify, my problem isn't when `sorts` has `null` value but when it doesn't. It has one item and the problem is that the list is written directly to JSON object without a property name first. My mistake I forgot to add exception, here it is:

```java
com.fasterxml.jackson.core.JsonGenerationException: Can not start an array, expecting field name
	at com.fasterxml.jackson.core.JsonGenerator._reportError(JsonGenerator.java:1872)
	at com.fasterxml.jackson.core.json.UTF8JsonGenerator._verifyValueWrite(UTF8JsonGenerator.java:957)
	at com.fasterxml.jackson.core.json.UTF8JsonGenerator.writeStartArray(UTF8JsonGenerator.java:281)
	at org.immutables.gson.stream.JsonGeneratorWriter.beginArray(JsonGeneratorWriter.java:62)
	at <my package>.entities.GsonAdaptersListContext$ListContextTypeAdapter.writeListContext(GsonAdaptersListContext.java:108)
...
```
...and line 108 is `out.beginArray()`.

**Comment by pogo19** [comment]
OK, I tried 2.2.8 now and it is already fixed:
```java
...
@Nullable List<Sort> sortsElements = instance.getSorts();
      if (sortsElements != null) {
        out.name("sorts");
        out.beginArray();
        for (Sort e : sortsElements) {
          sortsTypeAdapter.write(out, e);
        }
        out.endArray();
      } else if (out.getSerializeNulls()) {
        out.name("sorts");
        out.nullValue();
      }
...
```

**Comment by elucash** [comment]
Great! yes this was fixed in 2.2.6. Feel free to report any other bugs you might encounter
